eth-contracts/test/delegateManager.test.js

```diff
@@ -998,9 +990,95 @@ contract('DelegateManager', async (accounts) => {
       currentDuration = await delegateManager.getUndelegateLockupDuration()
       assert.isTrue(currentDuration.eq(newDuration))
     })
+
+    it('maximum delegators', async () => {
+      // Update max delegators to 5
+      let maxDelegators = 4
+      await mockGovernance.updateMaxDelegators(maxDelegators)
+      assert.equal(
+        fromBn(await delegateManager.getMaxDelegators()),
+        maxDelegators,
+        'Max delegators not updated'
+      )
+      let delegateAccountOffset = 4
+      let delegatorAccounts = accounts.slice(delegateAccountOffset, delegateAccountOffset + (maxDelegators + 1))
+      let singleDelegateAmount = toWei(10)
+      for (var i = 0; i < delegatorAccounts.length; i++) {
+        let delegator = delegatorAccounts[i]
+        // Transfer 1000 tokens to each delegator
+        await token.transfer(delegator, singleDelegateAmount, { from: treasuryAddress })
+        // Approve staking transfer
+        await token.approve(stakingAddress, singleDelegateAmount, { from: delegator })
+        if (i === (delegatorAccounts.length - 1)) {
+          await _lib.assertRevert(
+            delegateManager.delegateStake(
+              stakerAccount,
+              singleDelegateAmount,
+              { from: delegator }),
+            'Maximum delegators exceeded'
+          )
+        } else {
+          await delegateManager.delegateStake(
+            stakerAccount,
+            singleDelegateAmount,
+            { from: delegator })
+        }
+      }
+    })
+
+    it('min delegate stake', async () => {
+      // Update min delegation level configuration
+      let minDelegateStake = toWei(100)
+      await mockGovernance.updateMinDelegationAmount(minDelegateStake)
+      assert.isTrue(
+        minDelegateStake.eq(await delegateManager.getMinDelegationAmount()),
+        'Min delegation not updated')
+
+      // Approve staking transfer
+      await token.approve(
+        stakingAddress,
+        minDelegateStake,
+        { from: delegatorAccount1 })
+
+      // Confirm failure below minimum
+      await _lib.assertRevert(
+        delegateManager.delegateStake(
+          stakerAccount,
+          minDelegateStake.sub(toWei(10)),
+          { from: delegatorAccount1 }),
+        'Minimum delegation amount')
+
+      // Delegate min
+      await delegateManager.delegateStake(
+        stakerAccount,
+        minDelegateStake,
+        { from: delegatorAccount1 })
+
+      // Submit request to undelegate stake that will fail as total amount results in < min amount
+      let failUndelegateAmount = minDelegateStake.sub(toWei(30))
+      await delegateManager.requestUndelegateStake(stakerAccount, failUndelegateAmount, { from: delegatorAccount1 })
+      let undelegateRequestInfo = await delegateManager.getPendingUndelegateRequest(delegatorAccount1)
+      await _lib.advanceToTargetBlock(fromBn(undelegateRequestInfo.lockupExpiryBlock), web3)
+      await _lib.assertRevert(
```

Comment:
what is the reasoning for failing at the undelegate call instead of the requestUndelegate call? could you not just check mindelegationamount inside the request and fail then?

Reply:
good point, i figured that since you are allowed to still delegate more stake after requesting undelegation so it made sense to enforce when the operation actually happens. wydt?

Reply:
i see your point there but it logically seems unnecessary. like if you're going to requestUndelegate to below min and then delegateStake back to above min, like, why would you do that and why is allowing that better than just blocking the original requestUndelegate?